### PR TITLE
fix: Get shard id from a constant (topic, shard_index) to avoid error of shard_id FK violation

### PR DESCRIPTION
**Pull request by NGA-TRAN**
Closes https://github.com/influxdata/idpe/issues/16978

- [ ] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [ ] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

**Comment by pauldix** [comment]
I believe you need to use the shard index and lookup the ID. You can see an example of this in ingester2

**Comment by NGA-TRAN** [comment]
> I believe you need to use the shard index and lookup the ID. You can see an example of this in ingester2

Oh, I thought we are not supposed to use it and just hard-code. Anyway, will look into how Ingester2 works and do the same later

**Comment by NGA-TRAN** [comment]
OK, found `TRANSITION_SHARD_INDEX` in the Ingester2 which is a constant of 1 and we will read the `shard_id` of that constant index from the catalog only once when the Compactor is initiated. If I have time tomorrow/Sunday, I will do it. Otherwise, will do it on Monday.

Some other thought though. If we decide we will delete this `shard_id`, we can simply remove this FK constraint and do not need to worry about this half-hardcode (shard index = 1) and catalog read solution.

**Comment by pauldix** [comment]
Yeah, once we get everything moved over to Kafkaless, we can just remove all the shard references everywhere.

File: ioxd_compactor2/src/lib.rs [hunk 1]
```diff
@@ -22,6 +22,10 @@ use std::{
 use tokio_util::sync::CancellationToken;
 use trace::TraceCollector;
 
+// There is only one shard with index 1
+const TOPIC: &str = "iox-shared";
+const SHARD_INDEX: i32 = 1;
```
Comment (NGA-TRAN) [inline ioxd_compactor2/src/lib.rs:27]: In order to get shard ID, we need both topic and index. Even though we will remove them eventually, I do not want to break this rule and just follow Ingester2 to still use both topic and shard_index to get the shard id

File: compactor2/src/config.rs [hunk 2]
```diff
@@ -54,3 +70,41 @@ pub struct Config {
...
+            time_provider,
+            backoff_config,
+            partition_concurrency,
+            partition_minute_threshold,
+            max_desired_file_size_bytes,
+            percentage_max_file_size,
+            split_percentage,
+            shard_info: ShardInfo {
+                topic,
+                shard_index: ShardIndex::new(shard_index),
```
Comment (NGA-TRAN) [inline compactor2/src/config.rs:106]: Create this `new` function for us to pass in `shard_index` as an `i32` so we do not need to add `data_types` dependency into the crate `ioxd_compactor2`

**Review by NGA-TRAN — COMMENTED** [review]
OK. Here is the PR to read a shard id for a give and constant (topic, shard_index)

**Comment by pauldix** [comment]
@NGA-TRAN The shard ID doesn't change. It should be read once on initial startup of the compactor and then passed in as an initialization variable to the compactor logic (assuming there's some overall compactor server that can hold this variable)

**Comment by NGA-TRAN** [comment]
OK I make the data log shard reading only once at the compactor initiation time per Paul suggested. My comment above for @crepererum about caching it is no longer valid

**Review by pauldix — APPROVED** [review]

**Pull request merged**